trie/trie.go

```diff
@@ -602,7 +620,10 @@ func (t *Trie) Commit(onleaf LeafCallback) (common.Hash, int, error) {
 	// Do a quick check if we really need to commit, before we spin
 	// up goroutines. This can happen e.g. if we load a trie for reading storage
 	// values, but don't write to it.
-	if _, dirty := t.root.cache(); !dirty {
+	if hashedNode, dirty := t.root.cache(); !dirty {
+		// Replace the root node with the origin hash in order to
+		// ensure all resolved nodes are dropped after the commit.
+		t.root = hashedNode
```

Comment:
@holiman This can be a notable change. If the trie is clean and nothing-dirty-to-commit, the root node will be replaced with a hash.

The reason for adding this change is: I want to keep the guarantee that after commit operation, all the loaded/resolved nodes are released and converted to hash(except the embedded nodes). 

This guarantee is important for "pre-value-tracer". Whenever we load the node from the database, "pre-value-tracer" will track the value in memory and which can be used as "pre-value". However, tracer will be reset in every commit operation, so we have to ensure there is no resolved node anymore after commit as well.

Reply:
:+1: